### ISSUE TITLE
Direct processing of EBNF (for better diagnostics)

### ISSUE THREAD

**Issue by ArsenShnurkov** [patrickhuber/Pliant#81]
2006, Sheng-Jun Wang, Cheng-Zhi Jin, An extension of earley's algorithm for extended grammars
    https://link.springer.com/chapter/10.1007/978-1-4020-3953-9_22
    EUR 24.95 (I don't have this .pdf)

I don't know do you use it, or not. That's my question and possible proposition (to use).

**Comment by patrickhuber** [comment]
I purchased the paper a while back, but didn't spend time working on the implementation. 

If I remember correctly, the ebnf syntax is currently faked by generating node names during the grammar creation. If you specify a repetition

```
A := A { A }
```

It gets turned into two grammar rules

```
A := A {A}
{A} := λ | A {A}
```

This is the same for any grouping, though different groupings ex: [optional] or (group) have different rules that are generated based on the rules for ebnf to bnf. 

```
A:= A [B]
B:= 'b'
```

results in 

```
A := A [B]
[B] := λ | B
B := 'b'
```

This all happens in the (ebnf grammar generator)[https://github.com/patrickhuber/Pliant/blob/master/libraries/Pliant/Ebnf/EbnfGrammarGenerator.cs]. See the Grouping, Optional and Repetition methods. You can see its use in the [tests](https://github.com/patrickhuber/Pliant/blob/master/tests/Pliant.Tests.Unit/Ebnf/EbnfGrammarGeneratorTests.cs)

**Comment by ArsenShnurkov** [comment]
> This all happens in the ebnf grammar generator.

Yes, but there is no mechanism to trace errors back to rules in original EBNF grammar text.

**Comment by patrickhuber** [comment]
That is true, you would have to unwind the computed grammar rule post parse during AST evaluation. I do something similar with a single pass disambiguation algorithm. 

Another work around is to rewrite the extended functions out of the grammar. 

Found the paper, here is the table of operators listed:

| Operator   |  Name    |  
|----|------|
| a     | Terminal       | 
| A    | Non Terminal |
| E*   | Zero or More |
| E% | One or More|
| E<sub>1</sub> \| E<sub>2</sub> | Or |
| E<sub>1</sub>#E<sub>2</sub> | List of E<sub>1</sub>  seperated by E<sub>2</sub> |
| E<sub>1</sub> + E<sub>2</sub> | E<sub>1</sub> or E<sub>2</sub> or E<sub>1</sub>E<sub>2</sub>
| {E} | Group elements | 
| [E] | Zero or One | 
| E<sub>1</sub>&#9675;E<sub>2</sub> | Sequence of E<sub>1</sub> and E<sub>2</sub> or E<sub>1</sub> E<sub>2</sub> for short |

The paper suggests creating a state machine to represent the progress through each extended earley item. Currently we use the DottedRule class to represent the state of the parse so that would need to change. DottedRule is a linear subset of a finite state machine currently, though replacing it would ripple through the code base. 

It looks like the other change is adding the "Extend" function to the list of current (Predict, Complete and Scan). This function appears to compete with the "Complete" function for items. I'd like to see a sample implementation to see how they handle the full algorithm and parse nodes. The paper is pretty high level in detail.
